File: app/src/main/java/com/jmalinen/blackjack/ui/screens/GameScreen.kt
package com.jmalinen.blackjack.ui.screens

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import com.jmalinen.blackjack.engine.BasicStrategyAdvisor
import com.jmalinen.blackjack.model.GamePhase
import com.jmalinen.blackjack.model.GameState
import com.jmalinen.blackjack.model.PlayerAction
import com.jmalinen.blackjack.model.SurrenderPolicy
import com.jmalinen.blackjack.ui.components.ActionBar
import com.jmalinen.blackjack.ui.components.BetSelector
import com.jmalinen.blackjack.ui.components.DealerArea
import com.jmalinen.blackjack.ui.components.ExtraPlayersArea
import com.jmalinen.blackjack.ui.components.GameInfoBar
import com.jmalinen.blackjack.ui.components.InsuranceBar
import com.jmalinen.blackjack.ui.components.PlayerArea
import com.jmalinen.blackjack.ui.components.StrategyChartOverlay
import com.jmalinen.blackjack.ui.theme.FeltGreen
import com.jmalinen.blackjack.ui.theme.GoldAccent
import com.jmalinen.blackjack.viewmodel.GameViewModel

@Composable
fun GameScreen(
    viewModel: GameViewModel,
    onNavigateToSettings: () -> Unit
) {
    val state by viewModel.state.collectAsStateWithLifecycle()
    var showChart by remember { mutableStateOf(false) }
    val chartData = remember(state.rules) { BasicStrategyAdvisor.getChartData(state.rules) }

    Box(modifier = Modifier.fillMaxSize()) {
        GameScreenContent(
            state = state,
            onBetChanged = viewModel::adjustBet,
            onDeal = viewModel::deal,
            onHit = viewModel::hit,
            onStand = viewModel::stand,
            onDouble = viewModel::doubleDown,
            onSplit = viewModel::split,
            onSurrender = viewModel::surrender,
            onInsurance = viewModel::takeInsurance,
            onDeclineInsurance = viewModel::declineInsurance,
            onEvenMoney = viewModel::takeEvenMoney,
            onDeclineEvenMoney = viewModel::declineEvenMoney,
            onNewRound = viewModel::newRound,
            onReset = viewModel::resetGame,
            onToggleCoach = viewModel::toggleCoach,
            onToggleDeviations = viewModel::toggleDeviations,
            onToggleCount = viewModel::toggleCount,
            onShowChart = { showChart = true },
            onEnd = onNavigateToSettings
        )

        StrategyChartOverlay(
            visible = showChart,
            chartData = chartData,
            hasSurrender = state.rules.surrenderPolicy != SurrenderPolicy.NONE,
            onDismiss = { showChart = false }
        )
    }
}

@Composable
internal fun GameScreenContent(
    state: GameState,
    onBetChanged: (Int) -> Unit = {},
    onDeal: () -> Unit = {},
    onHit: () -> Unit = {},
    onStand: () -> Unit = {},
    onDouble: () -> Unit = {},
    onSplit: () -> Unit = {},
    onSurrender: () -> Unit = {},
    onInsurance: () -> Unit = {},
    onDeclineInsurance: () -> Unit = {},
    onEvenMoney: () -> Unit = {},
    onDeclineEvenMoney: () -> Unit = {},
    onNewRound: () -> Unit = {},
    onReset: () -> Unit = {},
    onToggleCoach: () -> Unit = {},
    onToggleDeviations: () -> Unit = {},
    onToggleCount: () -> Unit = {},
    onShowChart: () -> Unit = {},
    onEnd: () -> Unit = {}
) {
    Column(
        modifier = Modifier
            .fillMaxSize()
            .background(FeltGreen)
    ) {
        GameInfoBar(
            chips = state.chips,
            currentBet = state.currentBet,
            shoePenetration = state.shoePenetration,
            handsPlayed = state.handsPlayed,
            handsWon = state.handsWon,
            coachEnabled = state.coachEnabled,
            onToggleCoach = onToggleCoach,
            deviationsEnabled = state.deviationsEnabled,
            onToggleDeviations = onToggleDeviations,
            showCount = state.showCount,
            runningCount = state.runningCount,
            trueCount = state.trueCount,
            onToggleCount = onToggleCount,
            onShowChart = onShowChart,
            onEnd = onEnd
        )

        val hasExtraPlayers = state.extraPlayers.isNotEmpty()

        DealerArea(
            hand = state.dealerHand,
            showHoleCard = state.showDealerHoleCard,
            modifier = Modifier.weight(if (hasExtraPlayers) 0.7f else 1f),
            compact = hasExtraPlayers
        )

        if (hasExtraPlayers) {
            ExtraPlayersArea(
                extraPlayers = state.extraPlayers
            )
        }

        PlayerArea(
            hands = state.playerHands,
            activeHandIndex = state.activeHandIndex,
            handResults = state.handResults,
            phase = state.phase,
            currentBet = state.currentBet,
            modifier = Modifier.weight(if (hasExtraPlayers) 1.5f else 1.2f),
            cardScale = if (hasExtraPlayers) 0.85f else 1f
        )

        // Bottom action area — fixed height so card areas above don't shift
        Box(
            modifier = Modifier
                .fillMaxWidth()
                .height(160.dp),
            contentAlignment = Alignment.Center
        ) {
            when (state.phase) {
                GamePhase.BETTING -> {
                    BetSelector(
                        currentBet = state.currentBet,
                        chips = state.chips,
                        rules = state.rules,
                        onBetChanged = onBetChanged,
                        onDeal = onDeal
                    )
                }

                GamePhase.INSURANCE_OFFERED -> {
                    Column(
                        modifier = Modifier
                            .fillMaxWidth()
                            .padding(8.dp),
                        horizontalAlignment = Alignment.CenterHorizontally
                    ) {
                        Text(
                            text = "Insurance?",
                            color = Color.White,
                            fontSize = 16.sp,
                            fontWeight = FontWeight.Bold
                        )
                        InsuranceBar(
                            availableActions = state.availableActions,
                            onInsurance = onInsurance,
                            onDecline = onDeclineInsurance,
                            onEvenMoney = onEvenMoney,
                            onDeclineEvenMoney = onDeclineEvenMoney
                        )
                    }
                }

                GamePhase.PLAYER_TURN -> {
                    ActionBar(
                        availableActions = state.availableActions,
                        onHit = onHit,
                        onStand = onStand,
                        onDouble = onDouble,
                        onSplit = onSplit,
                        onSurrender = onSurrender
                    )
                }

                GamePhase.DEALING, GamePhase.DEALER_PEEK, GamePhase.EXTRA_PLAYERS_TURN, GamePhase.DEALER_TURN -> {
                    Text(
                        text = when (state.phase) {
                            GamePhase.DEALER_TURN -> "Dealer's turn..."
                            GamePhase.EXTRA_PLAYERS_TURN -> "Other players..."
                            else -> "Dealing..."
                        },
                        color = Color.White.copy(alpha = 0.7f),
                        fontSize = 16.sp
                    )
                }

                GamePhase.ROUND_COMPLETE -> {
                    ResultArea(
                        message = state.roundMessage,
                        payout = state.roundPayout,
                        onNewRound = onNewRound
                    )
                }

                GamePhase.GAME_OVER -> {
                    GameOverArea(
                        handsPlayed = state.handsPlayed,
                        handsWon = state.handsWon,
                        onReset = onReset
                    )
                }
            }
        }

        if (state.coachEnabled && state.coachFeedback.isNotEmpty()) {
            val isCorrect = state.coachFeedback.startsWith("Correct")
            Row(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(horizontal = 16.dp, vertical = 4.dp),
                horizontalArrangement = Arrangement.Center,
                verticalAlignment = Alignment.CenterVertically
            ) {
                Text(
                    text = state.coachFeedback,
                    color = if (isCorrect) Color(0xFF4CAF50) else Color(0xFFFF9800),
                    fontSize = 14.sp,
                    fontWeight = FontWeight.Bold,
                    textAlign = TextAlign.Center,
                    modifier = Modifier.weight(1f)
                )
                if (state.coachTotal > 0) {
                    Text(
                        text = "${state.coachCorrect}/${state.coachTotal}",
                        color = Color.White.copy(alpha = 0.6f),
                        fontSize = 13.sp,
                        fontWeight = FontWeight.Bold
                    )
                }
            }
        }
    }
}

@Composable
private fun ResultArea(
    message: String,
    payout: Int,
    onNewRound: () -> Unit
) {
    Column(
        modifier = Modifier
            .fillMaxWidth()
            .padding(16.dp),
        horizontalAlignment = Alignment.CenterHorizontally
    ) {
        Text(
            text = message,
            color = Color.White,
            fontSize = 20.sp,
            fontWeight = FontWeight.Bold,
            textAlign = TextAlign.Center
        )

        Spacer(modifier = Modifier.height(12.dp))

        Button(
            onClick = onNewRound,
            colors = ButtonDefaults.buttonColors(
                containerColor = GoldAccent,
                contentColor = Color.Black
            ),
            shape = RoundedCornerShape(24.dp)
        ) {
            Text("Next Hand", fontWeight = FontWeight.Bold, fontSize = 16.sp)
        }
    }
}

@Composable
private fun GameOverArea(
    handsPlayed: Int,
    handsWon: Int,
    onReset: () -> Unit
) {
    Column(
        modifier = Modifier
            .fillMaxWidth()
            .padding(16.dp),
        horizontalAlignment = Alignment.CenterHorizontally
    ) {
        Text(
            text = "Game Over!",
            color = Color(0xFFEF5350),
            fontSize = 24.sp,
            fontWeight = FontWeight.Bold
        )
        Text(
            text = "You're out of chips",
            color = Color.White.copy(alpha = 0.7f),
            fontSize = 14.sp
        )
        Text(
            text = "Played: $handsPlayed  Won: $handsWon",
            color = Color.White.copy(alpha = 0.7f),
            fontSize = 13.sp
        )

        Spacer(modifier = Modifier.height(12.dp))

        Button(
            onClick = onReset,
            colors = ButtonDefaults.buttonColors(
                containerColor = GoldAccent,
                contentColor = Color.Black
            )
        ) {
            Text("Play Again", fontWeight = FontWeight.Bold)
        }
    }
}

File: ios/Blackjack/Blackjack/Views/Screens/GameScreen.swift
import SwiftUI

struct GameScreen: View {
    let viewModel: GameViewModel
    let onNavigateToSettings: () -> Void

    @State private var showChart = false

    var body: some View {
        let state = viewModel.state
        let chartData = BasicStrategyAdvisor.getChartData(rules: state.rules)

        ZStack {
            GameScreenContent(
                state: state,
                onBetChanged: viewModel.adjustBet,
                onDeal: viewModel.deal,
                onHit: viewModel.hit,
                onStand: viewModel.stand,
                onDouble: viewModel.doubleDown,
                onSplit: viewModel.split,
                onSurrender: viewModel.surrender,
                onInsurance: viewModel.takeInsurance,
                onDeclineInsurance: viewModel.declineInsurance,
                onEvenMoney: viewModel.takeEvenMoney,
                onDeclineEvenMoney: viewModel.declineEvenMoney,
                onNewRound: viewModel.newRound,
                onReset: viewModel.resetGame,
                onToggleCoach: viewModel.toggleCoach,
                onToggleDeviations: viewModel.toggleDeviations,
                onToggleCount: viewModel.toggleCount,
                onShowChart: { showChart = true },
                onEnd: onNavigateToSettings
            )

            StrategyChartOverlay(
                visible: showChart,
                chartData: chartData,
                hasSurrender: state.rules.surrenderPolicy != .none,
                onDismiss: { showChart = false }
            )
        }
        .navigationBarHidden(true)
    }
}

private struct UiScales {
    let dealerScale: CGFloat
    let playerScale: CGFloat
    let extraPlayerScale: CGFloat
    let actionBoxHeight: CGFloat
    let compactActions: Bool

    init(dealerScale: CGFloat, playerScale: CGFloat, extraPlayerScale: CGFloat, actionBoxHeight: CGFloat, compactActions: Bool = false) {
        self.dealerScale = dealerScale
        self.playerScale = playerScale
        self.extraPlayerScale = extraPlayerScale
        self.actionBoxHeight = actionBoxHeight
        self.compactActions = compactActions
    }
}

private func computeScales(width: CGFloat, hasExtraPlayers: Bool) -> UiScales {
    if width < 380 {
        return UiScales(
            dealerScale: 0.65,
            playerScale: hasExtraPlayers ? 0.50 : 0.65,
            extraPlayerScale: 0.50,
            actionBoxHeight: 120,
            compactActions: true
        )
    } else if width < 600 {
        return UiScales(
            dealerScale: 1.0,
            playerScale: hasExtraPlayers ? 0.75 : 1.0,
            extraPlayerScale: 0.65,
            actionBoxHeight: 160
        )
    } else {
        return UiScales(
            dealerScale: 1.4,
            playerScale: hasExtraPlayers ? 1.19 : 1.4,
            extraPlayerScale: 1.05,
            actionBoxHeight: 200
        )
    }
}

private struct GameScreenContent: View {
    let state: GameState
    var onBetChanged: (Int) -> Void = { _ in }
    var onDeal: () -> Void = {}
    var onHit: () -> Void = {}
    var onStand: () -> Void = {}
    var onDouble: () -> Void = {}
    var onSplit: () -> Void = {}
    var onSurrender: () -> Void = {}
    var onInsurance: () -> Void = {}
    var onDeclineInsurance: () -> Void = {}
    var onEvenMoney: () -> Void = {}
    var onDeclineEvenMoney: () -> Void = {}
    var onNewRound: () -> Void = {}
    var onReset: () -> Void = {}
    var onToggleCoach: () -> Void = {}
    var onToggleDeviations: () -> Void = {}
    var onToggleCount: () -> Void = {}
    var onShowChart: () -> Void = {}
    var onEnd: () -> Void = {}

    var body: some View {
        GeometryReader { geometry in
            let hasExtraPlayers = !state.extraPlayers.isEmpty
            let scales = computeScales(width: geometry.size.width, hasExtraPlayers: hasExtraPlayers)

            VStack(spacing: 0) {
                GameInfoBar(
                    chips: state.chips,
                    shoePenetration: state.shoePenetration,
                    handsPlayed: state.handsPlayed,
                    handsWon: state.handsWon,
                    coachEnabled: state.coachEnabled,
                    onToggleCoach: onToggleCoach,
                    deviationsEnabled: state.deviationsEnabled,
                    onToggleDeviations: onToggleDeviations,
                    showCount: state.showCount,
                    runningCount: state.runningCount,
                    trueCount: state.trueCount,
                    onToggleCount: onToggleCount,
                    onShowChart: onShowChart,
                    onEnd: onEnd,
                    coachFeedback: state.coachFeedback,
                    coachCorrect: state.coachCorrect,
                    coachTotal: state.coachTotal
                )

                DealerArea(
                    hand: state.dealerHand,
                    showHoleCard: state.showDealerHoleCard,
                    compact: hasExtraPlayers,
                    cardScale: scales.dealerScale
                )

                if hasExtraPlayers {
                    ExtraPlayersArea(
                        extraPlayers: state.extraPlayers,
                        cardScale: scales.extraPlayerScale
                    )
                }

                PlayerArea(
                    hands: state.playerHands,
                    activeHandIndex: state.activeHandIndex,
                    handResults: state.handResults,
                    phase: state.phase,
                    currentBet: state.currentBet,
                    cardScale: scales.playerScale
                )

                // Bottom action area — fixed height
                ZStack {
                    switch state.phase {
                    case .betting:
                        BetSelector(
                            currentBet: state.currentBet,
                            chips: state.chips,
                            rules: state.rules,
                            onBetChanged: onBetChanged,
                            onDeal: onDeal
                        )

                    case .insuranceOffered:
                        VStack(spacing: 4) {
                            Text("Insurance?")
                                .font(.system(size: 16, weight: .bold))
                                .foregroundColor(.white)
                            InsuranceBar(
                                availableActions: state.availableActions,
                                onInsurance: onInsurance,
                                onDecline: onDeclineInsurance,
                                onEvenMoney: onEvenMoney,
                                onDeclineEvenMoney: onDeclineEvenMoney
                            )
                        }
                        .padding(8)

                    case .playerTurn:
                        ActionBar(
                            availableActions: state.availableActions,
                            onHit: onHit,
                            onStand: onStand,
                            onDouble: onDouble,
                            onSplit: onSplit,
                            onSurrender: onSurrender
                        )

                    case .dealing, .dealerPeek, .extraPlayersTurn, .dealerTurn:
                        Text(phaseMessage(state.phase))
                            .font(.system(size: 16))
                            .foregroundColor(.white.opacity(0.7))

                    case .roundComplete:
                        ResultArea(
                            message: state.roundMessage,
                            payout: state.roundPayout,
                            onNewRound: onNewRound,
                            compact: scales.compactActions
                        )

                    case .gameOver:
                        GameOverArea(
                            handsPlayed: state.handsPlayed,
                            handsWon: state.handsWon,
                            onReset: onReset,
                            compact: scales.compactActions
                        )
                    }
                }
                .frame(height: scales.actionBoxHeight)
                .frame(maxWidth: .infinity)
            }
            .background(Color.feltGreen)
        }
    }

    private func phaseMessage(_ phase: GamePhase) -> String {
        switch phase {
        case .dealerTurn: "Dealer's turn..."
        case .extraPlayersTurn: "Other players..."
        default: "Dealing..."
        }
    }
}

private struct ResultArea: View {
    let message: String
    let payout: Int
    let onNewRound: () -> Void
    var compact: Bool = false

    var body: some View {
        VStack(spacing: compact ? 4 : 12) {
            Text(message)
                .font(.system(size: compact ? 16 : 20, weight: .bold))
                .foregroundColor(.white)
                .multilineTextAlignment(.center)

            Button(action: onNewRound) {
                Text("Next Hand")
                    .font(.system(size: compact ? 14 : 16, weight: .bold))
                    .foregroundColor(.black)
                    .padding(.horizontal, 24)
                    .padding(.vertical, 10)
                    .background(Color.goldAccent)
                    .clipShape(Capsule())
            }
        }
        .padding(compact ? 8 : 16)
    }
}

private struct GameOverArea: View {
    let handsPlayed: Int
    let handsWon: Int
    let onReset: () -> Void
    var compact: Bool = false

    var body: some View {
        VStack(spacing: compact ? 2 : 12) {
            Text("Game Over!")
                .font(.system(size: compact ? 18 : 24, weight: .bold))
                .foregroundColor(Color(hex: 0xEF5350))

            if !compact {
                Text("You're out of chips")
                    .font(.system(size: 14))
                    .foregroundColor(.white.opacity(0.7))
            }

            Text(compact
                ? "Out of chips · Played: \(handsPlayed)  Won: \(handsWon)"
                : "Played: \(handsPlayed)  Won: \(handsWon)")
                .font(.system(size: compact ? 11 : 13))
                .foregroundColor(.white.opacity(0.7))

            Button(action: onReset) {
                Text("Play Again")
                    .font(.system(size: compact ? 13 : 14, weight: .bold))
                    .foregroundColor(.black)
                    .padding(.horizontal, 20)
                    .padding(.vertical, 8)
                    .background(Color.goldAccent)
                    .cornerRadius(8)
            }
        }
        .padding(compact ? 4 : 16)
    }
}
